server/src/main/java/org/apache/cloudstack/diagnostics/DiagnosticsServiceImpl.java

```diff
@@ -0,0 +1,103 @@
+import org.apache.cloudstack.api.command.admin.diagnostics.ExecuteDiagnosticsCmd;
+import org.apache.cloudstack.diangostics.DiagnosticsService;
+import org.apache.log4j.Logger;
+
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class DiagnosticsServiceImpl extends ManagerBase implements PluggableService, DiagnosticsService {
+    private static final Logger LOGGER = Logger.getLogger(DiagnosticsServiceImpl.class);
+
+    @Inject
+    private RouterControlHelper routerControlHelper;
+    @Inject
+    private AgentManager agentManager;
+    @Inject
+    private VMInstanceDao instanceDao;
+
+    @Override
+    public Map<String, String> runDiagnosticsCommand(final ExecuteDiagnosticsCmd cmd) throws AgentUnavailableException, InvalidParameterValueException {
+        final Long vmId = cmd.getId();
+        final String cmdType = cmd.getType().getValue();
+        final String cmdAddress = cmd.getAddress();
+        final String optionalArguments = cmd.getOptionalArguments();
+        final VMInstanceVO vmInstance = instanceDao.findByIdTypes(vmId, VirtualMachine.Type.ConsoleProxy, VirtualMachine.Type.DomainRouter, VirtualMachine.Type.SecondaryStorageVm);
+
+        if (vmInstance == null) {
+            LOGGER.error("Invalid system vm id provided " + vmId);
+            throw new InvalidParameterValueException("Unable to find a system vm with id " + vmId);
+        }
+        final Long hostId = vmInstance.getHostId();
+
+        if (hostId == null) {
+            LOGGER.warn("Unable to find host for virtual machine instance: " + vmInstance.getInstanceName());
+            throw new CloudRuntimeException("Unable to find host for virtual machine instance: " + vmInstance.getInstanceName());
+        }
+
+        final DiagnosticsCommand command = new DiagnosticsCommand(cmdType, cmdAddress, optionalArguments);
+        command.setAccessDetail(NetworkElementCommand.ROUTER_IP, routerControlHelper.getRouterControlIp(vmInstance.getId()));
+
+        Answer answer;
+
+        try {
+            answer = agentManager.send(hostId, command);
+            if (answer instanceof DiagnosticsAnswer) {
+                final Map<String, String> detailsMap = ((DiagnosticsAnswer) answer).getExecutionDetails();
+                if (!detailsMap.isEmpty()) {
+                    return detailsMap;
+                } else {
+                    LOGGER.error("Failed to parse diagnostics command execution results: " + answer.getDetails());
+                    throw new CloudRuntimeException("Failed to parse diagnostics command execution results ");
+                }
+
+            } else {
+                LOGGER.error("Failed to execute diagnostics command: " + answer.getDetails());
+                throw new CloudRuntimeException("Failed to execute diagnostics command: " + answer.getDetails());
+            }
+        } catch (OperationTimedoutException e) {
+            LOGGER.warn("Timed Out", e);
```

Comment:
a warning should be an error that can be handled. throwing and warning seem kind of contradictory